{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "708e5c42-4cbb-4970-8fc7-a78fcdca4269",
   "metadata": {},
   "source": [
    "# Data Wrangling test\n",
    "\n",
    "In this notebook, a data wrangling approach has been taken to process and integrate two types of biological data: RNA sequencing data and serum protein data, along with associated clinical information about patients. \n",
    "\n",
    "The initial steps involved reshaping and cleaning each dataset separately, ensuring that they were in a format conducive to analysis. This included standardizing measurement units, harmonizing column names, and appending relevant clinical details to each record. Following this, the two distinct datasets were merged into one comprehensive dataset, providing a unified view that combines information from RNA sequencing with protein data from serum samples. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70287c7e-f5ed-4cae-81cd-e8c4f858ef8b",
   "metadata": {},
   "source": [
    "# 1 - Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b0c599f5-63ac-40c5-9b62-42f684bd59fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fd738af5-36c2-441e-9d94-7637fe11b011",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_excel('TechnicalTest-DataWrangling-2.xlsx',sheet_name = None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e55c31a6-cc3b-429d-abd5-a2b8d6f9ac9f",
   "metadata": {},
   "source": [
    "# 2 - Formating Serum protein data\n",
    "\n",
    "The code block below performs a series of data wrangling operations on a DataFrame named **Serum**, which is a subset of a larger dataset focused on **Serum Protein data**. \n",
    "\n",
    "The initial step involves reshaping this DataFrame into a long format using the melt function. This transformation takes the 'Serum IL-6 (g/L)' and 'Serum IL-6 Receptor (mg/L)' columns and converts them into rows. \n",
    "\n",
    "In the newly transformed DataFrame, two new columns are created: 'Measurement', which holds the type of measurement (either 'IL-6' or 'IL-6 Receptor'), and 'Result', which stores the corresponding measurement values. The code then proceeds to extract and refine these measurement types and units. It assigns proper gene symbols ('IL-6' or 'IL6R' for 'IL-6 Receptor') and standardizes the result units (converting 'IL6R' results from mg/L to g/L for uniformity).\n",
    "\n",
    "Additionally, the 'Material type' column is added and set to 'SERUM' for all rows, indicating the type of material from which the measurements were taken. The 'Measurement' column is dropped as its information is now captured in other columns, leading to a cleaner and more concise DataFrame structure.\n",
    "\n",
    "Finally, the DataFrame is stored in a new variable, 'Serum_clean'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f30da90b-1292-47c5-8068-f9fee7aaee55",
   "metadata": {},
   "outputs": [],
   "source": [
    "Serum = data['Serum Protein data']\n",
    "\n",
    "# Melt the DataFrame to long format\n",
    "df_long = Serum.melt(id_vars=['Patient', 'Sample'], \n",
    "                  value_vars=['Serum IL-6 (g/L)', 'Serum IL-6 Receptor (mg/L)'],\n",
    "                  var_name='Measurement', value_name='Result')\n",
    "\n",
    "\n",
    "# Extract Gene Symbol and Result Units\n",
    "df_long['Gene_symbol'] = df_long['Measurement'].apply(lambda x: 'IL-6' if 'IL-6 (g/L)' in x else 'IL-6 Receptor')\n",
    "df_long['Result_Units'] = df_long['Measurement'].apply(lambda x: x.split(' ')[-1].strip('()'))\n",
    "# Update the 'Gene_symbol' column\n",
    "df_long['Gene_symbol'] = df_long['Gene_symbol'].replace({'IL-6 Receptor': 'IL6R'})\n",
    "\n",
    "\n",
    "# Set the Material type as 'SERUM'\n",
    "df_long['Material type'] = 'SERUM'\n",
    "\n",
    "# Drop the 'Measurement' column and rearrange the columns\n",
    "df_long = df_long.drop('Measurement', axis=1)\n",
    "df_long = df_long[['Patient', 'Sample', 'Material type', 'Gene_symbol', 'Result', 'Result_Units']]\n",
    "\n",
    "# Convert 'IL6R' results from mg/L to g/L\n",
    "df_long.loc[df_long['Gene_symbol'] == 'IL6R', 'Result'] = df_long['Result'].apply(pd.to_numeric, errors='coerce') * 0.001\n",
    "# Update the 'Result_Units' for 'IL6R' to 'g/L'\n",
    "df_long.loc[df_long['Gene_symbol'] == 'IL6R', 'Result_Units'] = 'g/L'\n",
    "\n",
    "# Display the transformed DataFrame\n",
    "Serum_clean = df_long.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "dd8e6583-5d97-4c7e-b9ae-040290c90dc8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Material type</th>\n",
       "      <th>Gene_symbol</th>\n",
       "      <th>Result</th>\n",
       "      <th>Result_Units</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1001</td>\n",
       "      <td>S-CRC1-A</td>\n",
       "      <td>SERUM</td>\n",
       "      <td>IL-6</td>\n",
       "      <td>73.7</td>\n",
       "      <td>g/L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1001</td>\n",
       "      <td>S-CRC1-B</td>\n",
       "      <td>SERUM</td>\n",
       "      <td>IL-6</td>\n",
       "      <td>67.7</td>\n",
       "      <td>g/L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001</td>\n",
       "      <td>S-CRC1-C</td>\n",
       "      <td>SERUM</td>\n",
       "      <td>IL-6</td>\n",
       "      <td>71.3</td>\n",
       "      <td>g/L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1002</td>\n",
       "      <td>S-CRC2-A</td>\n",
       "      <td>SERUM</td>\n",
       "      <td>IL-6</td>\n",
       "      <td>QNS</td>\n",
       "      <td>g/L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1002</td>\n",
       "      <td>S-CRC2-B</td>\n",
       "      <td>SERUM</td>\n",
       "      <td>IL-6</td>\n",
       "      <td>33.1</td>\n",
       "      <td>g/L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Patient    Sample Material type Gene_symbol Result Result_Units\n",
       "0     1001  S-CRC1-A         SERUM        IL-6   73.7          g/L\n",
       "1     1001  S-CRC1-B         SERUM        IL-6   67.7          g/L\n",
       "2     1001  S-CRC1-C         SERUM        IL-6   71.3          g/L\n",
       "3     1002  S-CRC2-A         SERUM        IL-6    QNS          g/L\n",
       "4     1002  S-CRC2-B         SERUM        IL-6   33.1          g/L"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Serum_clean.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbc73f3e-6d49-405d-9d4d-35e3bda7707b",
   "metadata": {},
   "source": [
    "# 3 - Formatting RNA-seq data\n",
    "\n",
    "The code block below processes RNA sequencing data, initially restructuring it into a long format for ease of analysis. This transformation results in each row representing a specific gene's expression level in a particular sample. Additional details like measurement units and material type are appended to enhance clarity and consistency. The processed RNA data is then merged with tissue sample metadata, enriching the dataset with contextual information like patient numbers and sample types. To ensure uniformity in terminology, sample type descriptions are standardized through a mapping dictionary, aligning terms like 'Normal', 'Liver Tumor', and 'Metastic Lung' to 'NORMAL', 'PRIMARY', and 'METASTATIC', respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "e9f749a8-3901-4754-8c1a-ed12e87b677e",
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA = data['RNA-seq (RPKM)']\n",
    "\n",
    "RNA_clean = RNA.melt(id_vars='GeneID', var_name='Sample_ID', value_name='Result')\n",
    "RNA_clean['Result_Units'] = 'RPKM'\n",
    "RNA_clean['Material_type'] = 'RNA'\n",
    "RNA_clean.rename(columns = {'GeneID':'Gene_Symbol'},inplace = True)\n",
    "\n",
    "# Merging this data with Tissue Sample Metadata\n",
    "Tiss_metadata = data['Tissue Sample Metadata']\n",
    "Tiss_metadata = Tiss_metadata[['Patient  Number','Sample','Sample type']]\n",
    "\n",
    "RNA_clean_metadata = pd.merge(RNA_clean, Tiss_metadata, left_on = 'Sample_ID', right_on = 'Sample')\n",
    "\n",
    "# Replace values in the 'Sample' column\n",
    "\n",
    "replacements = {\n",
    "    'Normal': 'NORMAL',\n",
    "    'Liver Tumor': 'PRIMARY',\n",
    "    'Metastic Lung': 'METASTATIC'\n",
    "}\n",
    "\n",
    "# Using a loop to replace each key in the dictionary with its corresponding value\n",
    "for key, value in replacements.items():\n",
    "    RNA_clean_metadata['Sample type'] = RNA_clean_metadata['Sample type'].str.replace(key, value)\n",
    "\n",
    "RNA_clean_metadata.rename(columns = {'Sample type':'Sample_General_Pathology'},inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "1dd24ab5-2928-4226-a77f-773635fbf2de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Result</th>\n",
       "      <th>Result_Units</th>\n",
       "      <th>Material_type</th>\n",
       "      <th>Patient  Number</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Sample_General_Pathology</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ICAM1</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>4.662774</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>RNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>NORMAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IL6</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>RNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>NORMAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IL6R</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>12.153740</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>RNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>NORMAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VCAM1</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>4.596351</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>RNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>NORMAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SELE</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>0.077086</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>RNA</td>\n",
       "      <td>1001</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>NORMAL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene_Symbol Sample_ID     Result Result_Units Material_type  \\\n",
       "0       ICAM1     CRC1N   4.662774         RPKM           RNA   \n",
       "1         IL6     CRC1N   0.000000         RPKM           RNA   \n",
       "2        IL6R     CRC1N  12.153740         RPKM           RNA   \n",
       "3       VCAM1     CRC1N   4.596351         RPKM           RNA   \n",
       "4        SELE     CRC1N   0.077086         RPKM           RNA   \n",
       "\n",
       "   Patient  Number Sample Sample_General_Pathology  \n",
       "0             1001  CRC1N                   NORMAL  \n",
       "1             1001  CRC1N                   NORMAL  \n",
       "2             1001  CRC1N                   NORMAL  \n",
       "3             1001  CRC1N                   NORMAL  \n",
       "4             1001  CRC1N                   NORMAL  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RNA_clean_metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b1a1519-4992-4020-972e-92594de60ff9",
   "metadata": {},
   "source": [
    "# 4 - Merge data with Patient clinical data\n",
    "\n",
    "The code block below involves the integration and standardization of **RNA** and **serum** datasets with **patient clinical data**, followed by formatting and cleaning operations to prepare the datasets for analysis.\n",
    "\n",
    "**Merge RNA and Serum Data with Clinical Data:**\n",
    "The RNA dataset (RNA_clean_metadata) is merged with data['Patient_clinical_data'] based on the 'Patient Number' field. Similarly, the serum dataset (Serum_clean) is merged with the same clinical data. This integration enriches both datasets with additional clinical context for each patient, creating comprehensive datasets (RNA_full and Serum_full).\n",
    "\n",
    "**Rename Columns and Create Unique Identifiers:**\n",
    "In both merged datasets, the column names are standardized (e.g., renaming 'Patient Number' to 'Patient_ID'). This standardization aids in data consistency and readability.\n",
    "A new column, 'Unique_Patient_ID', is created by concatenating the 'Study_ID' and 'Patient_ID', providing a unique identifier for each patient across the datasets.\n",
    "\n",
    "**Standardize Sex Descriptions:**\n",
    "Both datasets have the 'Sex' column standardized, replacing abbreviations ('M' and 'F') with full descriptions ('MALE' and 'FEMALE'). This is achieved using a dictionary (replacements) and a loop that replaces each abbreviation with its corresponding full word. This step improves the clarity of the data.\n",
    "\n",
    "**Handle Duplicates and Finalize Data Structure:**\n",
    "Any duplicated columns in the RNA_full dataset are identified and dropped to avoid redundancy and potential errors during analysis.\n",
    "For the serum dataset, the columns are explicitly selected and reordered to ensure a clean and structured format. The final selection of columns includes identifiers, demographic information, and key measurement data.\n",
    "\n",
    "## 4.1 - RNA-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "66599c75-2908-4c45-9d9e-f512b5606d84",
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA_full = pd.merge(RNA_clean_metadata, data['Patient_clinical_data'], on = 'Patient  Number')\n",
    "# rename Patient Number column\n",
    "RNA_full.rename(columns = {'Patient  Number':'Patient_ID', 'Sample' : 'Sample_ID'},inplace = True)\n",
    "# Creating variable Unique_Patient_ID\n",
    "RNA_full['Unique_Patient_ID'] = RNA_full['Study_ID'] + '_' + RNA_full['Patient_ID'].astype(str)\n",
    "# Changing the entries for the column sex\n",
    "replacements = {\n",
    "    'M': 'MALE',\n",
    "    'F': 'FEMALE'\n",
    "}\n",
    "# Using a loop to replace each key in the dictionary with its corresponding value\n",
    "for key, value in replacements.items():\n",
    "    RNA_full['Sex'] = RNA_full['Sex'].str.replace(key, value)\n",
    "\n",
    "# Drop the duplicated column in RNA_full DataFrame\n",
    "RNA_full = RNA_full.loc[:, ~RNA_full.columns.duplicated()]\n",
    "RNA_full = RNA_full[['Study_ID','Patient_ID','Unique_Patient_ID','Sex','Age','Sample_ID','Sample_General_Pathology','Material_type','Gene_Symbol','Result','Result_Units']].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb7221e2-9504-42d7-993a-b7bd9caf5e14",
   "metadata": {},
   "source": [
    "## 4.2 - Serum data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "1a5a2f3b-57cf-4df0-868b-2278498eb2e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "Serum_full = pd.merge(Serum_clean, data['Patient_clinical_data'], left_on = 'Patient', right_on = 'Patient  Number')\n",
    "# rename Patient column\n",
    "Serum_full.rename(columns = {'Patient':'Patient_ID', 'Sample' : 'Sample_ID','Material type':'Material_type','Gene_symbol' : 'Gene_Symbol'},inplace = True)\n",
    "# Creating variable Unique_Patient_ID\n",
    "Serum_full['Unique_Patient_ID'] = Serum_full['Study_ID'] + '_' + Serum_full['Patient_ID'].astype(str)\n",
    "# Changing the entries for the column sex\n",
    "replacements = {\n",
    "    'M': 'MALE',\n",
    "    'F': 'FEMALE'\n",
    "}\n",
    "# Using a loop to replace each key in the dictionary with its corresponding value\n",
    "for key, value in replacements.items():\n",
    "    Serum_full['Sex'] = Serum_full['Sex'].str.replace(key, value)\n",
    "\n",
    "Serum_full = Serum_full[['Study_ID','Patient_ID','Unique_Patient_ID','Sex','Age','Sample_ID','Material_type','Gene_Symbol','Result','Result_Units']].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "967b9660-1855-4412-807e-969dff564244",
   "metadata": {},
   "source": [
    "# 5 - Concatenating both dataframes\n",
    "\n",
    "This code block effectively consolidates two distinct datasets, RNA_full and Serum_full, into a single comprehensive DataFrame. It achieves this by first resetting the indices of both DataFrames to ensure a smooth and error-free concatenation. Following this, the two datasets are merged into one unified DataFrame along the rows, creating a rich dataset that combines the detailed molecular data from both RNA and serum samples with their corresponding clinical data. Additionally, the code introduces a new 'Status' column initialized with missing values, leaving room for future updates or categorization based on upcoming analyses or criteria."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "598f0571-0de5-4e2b-b04b-2863c824120c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assuming RNA_full and Serum_full are your DataFrames\n",
    "RNA_full_reset = RNA_full.reset_index(drop=True)\n",
    "Serum_full_reset = Serum_full.reset_index(drop=True)\n",
    "\n",
    "# Concatenate the DataFrames\n",
    "final_df = pd.concat([RNA_full_reset, Serum_full_reset], axis=0) \n",
    "final_df['Status'] = pd.NA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ea4a655-9939-4a7e-895f-b42068540b14",
   "metadata": {},
   "source": [
    "# 6 - Write output\n",
    "\n",
    "Export the dataframe as a .csv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "f14e57d2-5422-417c-bf59-16729430090c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Unique_Patient_ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sample_General_Pathology</th>\n",
       "      <th>Material_type</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Result</th>\n",
       "      <th>Result_Units</th>\n",
       "      <th>Status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1001</td>\n",
       "      <td>RO4532A_1001</td>\n",
       "      <td>MALE</td>\n",
       "      <td>41.673240</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>NORMAL</td>\n",
       "      <td>RNA</td>\n",
       "      <td>ICAM1</td>\n",
       "      <td>4.662774</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>2310</td>\n",
       "      <td>RO4532A_2310</td>\n",
       "      <td>MALE</td>\n",
       "      <td>62.747454</td>\n",
       "      <td>CRC10M</td>\n",
       "      <td>METASTATIC</td>\n",
       "      <td>RNA</td>\n",
       "      <td>ICAM1</td>\n",
       "      <td>2.482575</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>2310</td>\n",
       "      <td>RO4532A_2310</td>\n",
       "      <td>MALE</td>\n",
       "      <td>62.747454</td>\n",
       "      <td>CRC10T</td>\n",
       "      <td>PRIMARY</td>\n",
       "      <td>RNA</td>\n",
       "      <td>ICAM1</td>\n",
       "      <td>10.449776</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1508</td>\n",
       "      <td>RO4532A_1508</td>\n",
       "      <td>MALE</td>\n",
       "      <td>77.947824</td>\n",
       "      <td>CRC8T</td>\n",
       "      <td>PRIMARY</td>\n",
       "      <td>RNA</td>\n",
       "      <td>ICAM1</td>\n",
       "      <td>6.433474</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1508</td>\n",
       "      <td>RO4532A_1508</td>\n",
       "      <td>MALE</td>\n",
       "      <td>77.947824</td>\n",
       "      <td>CRC8N</td>\n",
       "      <td>NORMAL</td>\n",
       "      <td>RNA</td>\n",
       "      <td>ICAM1</td>\n",
       "      <td>8.496029</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1003</td>\n",
       "      <td>RO4532A_1003</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>51.000000</td>\n",
       "      <td>CRC3T</td>\n",
       "      <td>PRIMARY</td>\n",
       "      <td>RNA</td>\n",
       "      <td>VCAM1</td>\n",
       "      <td>7.77177</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1004</td>\n",
       "      <td>RO4532A_1004</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>66.229214</td>\n",
       "      <td>CRC4T</td>\n",
       "      <td>PRIMARY</td>\n",
       "      <td>RNA</td>\n",
       "      <td>VCAM1</td>\n",
       "      <td>8.829485</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1003</td>\n",
       "      <td>RO4532A_1003</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>51.000000</td>\n",
       "      <td>CRC3M</td>\n",
       "      <td>METASTATIC</td>\n",
       "      <td>RNA</td>\n",
       "      <td>VCAM1</td>\n",
       "      <td>3.18751</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1002</td>\n",
       "      <td>RO4532A_1002</td>\n",
       "      <td>MALE</td>\n",
       "      <td>53.200569</td>\n",
       "      <td>CRC2N</td>\n",
       "      <td>NORMAL</td>\n",
       "      <td>RNA</td>\n",
       "      <td>VCAM1</td>\n",
       "      <td>4.827985</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1105</td>\n",
       "      <td>RO4532A_1105</td>\n",
       "      <td>MALE</td>\n",
       "      <td>53.304328</td>\n",
       "      <td>CRC5M</td>\n",
       "      <td>METASTATIC</td>\n",
       "      <td>RNA</td>\n",
       "      <td>VCAM1</td>\n",
       "      <td>4.299935</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>148 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Study_ID  Patient_ID Unique_Patient_ID     Sex        Age Sample_ID  \\\n",
       "0   RO4532A        1001      RO4532A_1001    MALE  41.673240     CRC1N   \n",
       "85  RO4532A        2310      RO4532A_2310    MALE  62.747454    CRC10M   \n",
       "95  RO4532A        2310      RO4532A_2310    MALE  62.747454    CRC10T   \n",
       "80  RO4532A        1508      RO4532A_1508    MALE  77.947824     CRC8T   \n",
       "75  RO4532A        1508      RO4532A_1508    MALE  77.947824     CRC8N   \n",
       "..      ...         ...               ...     ...        ...       ...   \n",
       "38  RO4532A        1003      RO4532A_1003  FEMALE  51.000000     CRC3T   \n",
       "53  RO4532A        1004      RO4532A_1004  FEMALE  66.229214     CRC4T   \n",
       "43  RO4532A        1003      RO4532A_1003  FEMALE  51.000000     CRC3M   \n",
       "18  RO4532A        1002      RO4532A_1002    MALE  53.200569     CRC2N   \n",
       "73  RO4532A        1105      RO4532A_1105    MALE  53.304328     CRC5M   \n",
       "\n",
       "   Sample_General_Pathology Material_type Gene_Symbol     Result Result_Units  \\\n",
       "0                    NORMAL           RNA       ICAM1   4.662774         RPKM   \n",
       "85               METASTATIC           RNA       ICAM1   2.482575         RPKM   \n",
       "95                  PRIMARY           RNA       ICAM1  10.449776         RPKM   \n",
       "80                  PRIMARY           RNA       ICAM1   6.433474         RPKM   \n",
       "75                   NORMAL           RNA       ICAM1   8.496029         RPKM   \n",
       "..                      ...           ...         ...        ...          ...   \n",
       "38                  PRIMARY           RNA       VCAM1    7.77177         RPKM   \n",
       "53                  PRIMARY           RNA       VCAM1   8.829485         RPKM   \n",
       "43               METASTATIC           RNA       VCAM1    3.18751         RPKM   \n",
       "18                   NORMAL           RNA       VCAM1   4.827985         RPKM   \n",
       "73               METASTATIC           RNA       VCAM1   4.299935         RPKM   \n",
       "\n",
       "   Status  \n",
       "0     NaN  \n",
       "85    NaN  \n",
       "95    NaN  \n",
       "80    NaN  \n",
       "75    NaN  \n",
       "..    ...  \n",
       "38    NaN  \n",
       "53    NaN  \n",
       "43    NaN  \n",
       "18    NaN  \n",
       "73    NaN  \n",
       "\n",
       "[148 rows x 12 columns]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df.sort_values(by='Gene_Symbol')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "759d7ce3-87d1-4daa-822e-ea83c07eae4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df.to_csv('DataWrangling_test_result.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "998ce870-e6f8-4c45-9bdb-74a2dc53e67d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "google_batch",
   "language": "python",
   "name": "google_batch"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
